synapse/storage/persist_events.py

```diff
@@ -734,7 +754,86 @@ async def _get_new_state_after_events(
             state_res_store=StateResolutionStore(self.main_store),
         )
 
-        return res.state, None
+        # If the returned state matches the state group of one of the new
+        # forward extremities then we check if we are able to prune some state
+        # extremities.
+        if res.state_group and res.state_group in new_state_groups:
```

Comment:
I'd prefer an early exit here than a huge if statement. (Or can we move the body of the `if` to a separate function?)

Reply:
... moving it out to a separate function might also mean we can do early exit from there, rather than the ungainly `drop = False; break` stuff here.

Reply:
That sounds much nicer. 👍